fe/fe-core/src/main/java/org/apache/doris/rewrite/ElementAtToSlotRefRule.java

```diff
@@ -0,0 +1,135 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.rewrite;
+
+import org.apache.doris.analysis.Analyzer;
+import org.apache.doris.analysis.Expr;
+import org.apache.doris.analysis.FunctionCallExpr;
+import org.apache.doris.analysis.LiteralExpr;
+import org.apache.doris.analysis.SlotRef;
+import org.apache.doris.catalog.Type;
+import org.apache.doris.common.AnalysisException;
+import org.apache.doris.rewrite.ExprRewriter.ClauseType;
+
+import com.google.common.collect.Lists;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Transform element_at function to SlotReference for variant sub-column access.
+ * This optimization will help query engine to prune as many sub columns as possible
+ * to speed up query.
+ * eg: element_at(element_at(v, "a"), "b") -> SlotReference(column=v, subColLabels=["a", "b"])
+ */
+
+public class ElementAtToSlotRefRule implements ExprRewriteRule  {
+    public static final ElementAtToSlotRefRule INSTANCE = new ElementAtToSlotRefRule();
+
+    @Override
+    public Expr apply(Expr expr, Analyzer analyzer, ClauseType clauseType) throws AnalysisException {
+        // Only check element at of variant all rewrited to slots
+        List<Expr> elementAtFunctions = Lists.newArrayList();
+        getElementAtFunction(expr, elementAtFunctions);
+        if (!elementAtFunctions.isEmpty()) {
+            throw new AnalysisException("element_at should not appear in common rewrite stage");
+        }
+        return expr;
+    }
+
+    private Expr apply(Expr expr, Analyzer analyzer) throws AnalysisException  {
+        if (!(isElementAtOfVariantType(expr))) {
+            return expr;
+        }
+        List<SlotRef> slotRefs =  Lists.newArrayList();
+        expr.collect(SlotRef.class, slotRefs);
+        SlotRef slot = slotRefs.get(0);
+        List<Expr> pathsExpr = Lists.newArrayList();
+        expr.collect(Expr::isLiteral, pathsExpr);
+        List<String> fullPaths = pathsExpr.stream()
+                .map(node -> ((LiteralExpr) node).getStringValue())
+                .collect(Collectors.toList());
+        slot.setSubColPath(fullPaths);
+        slot.analyzeImpl(analyzer);
+        return slot;
+    }
+
+    private static boolean isElementAtOfVariantType(Expr expr) {
+        if (!(expr instanceof FunctionCallExpr)) {
+            return false;
+        }
+        FunctionCallExpr functionCallExpr = (FunctionCallExpr) expr;
+        return functionCallExpr.getFnName().getFunction().equalsIgnoreCase("element_at")
+                && functionCallExpr.getType() == Type.VARIANT;
```

Comment:
functionCallExpr.getType() is return type of function. I think the type of the first arg should be checked instead.